h2o-extensions/target-encoder/src/main/java/ai/h2o/targetencoding/BroadcastJoinForTargetEncoder.java

```diff
@@ -0,0 +1,174 @@
+import water.fvec.Frame;
+
+class BroadcastJoinForTargetEncoder {
+
+  static class FrameWithEncodingDataToArray extends MRTask<FrameWithEncodingDataToArray> {
+    
+    /**
+     * Numerators and denominators are being stored in the same row-array of the 2D array.
+     * 
+     *        _encodingDataPerNode[k][0 .. numOfCatLevels -1] will be storing numerators
+     *        _encodingDataPerNode[k][numOfCatLevels .. (2*numOfCatLevels -1)] will be storing denominators,
+     *        
+     *        where k is a current fold value (folds = {0,1,2,3...k}) or 0 when _foldColumnIdx == -1; 
+     */
+    int[][] _encodingDataPerNode = null;
+    
+    int _categoricalColumnIdx, _foldColumnIdx, _numeratorIdx, _denominatorIdx, _cardinalityOfCatCol, _maxFoldValue;
+
+    FrameWithEncodingDataToArray(int categoricalColumnIdx, int foldColumnId, int numeratorIdx, int denominatorIdx, int cardinalityOfCatCol, int maxFoldValue) {
+      _categoricalColumnIdx = categoricalColumnIdx;
+      _foldColumnIdx = foldColumnId;
+      _numeratorIdx = numeratorIdx;
+      _denominatorIdx = denominatorIdx;
+      _cardinalityOfCatCol = cardinalityOfCatCol;
+
+      if(foldColumnId == -1) {
+        _encodingDataPerNode = new int[1][_cardinalityOfCatCol * 2];
+      } else {
+        assert maxFoldValue >= 1 : "There should be at leas two folds in the fold column";
+        assert _cardinalityOfCatCol > 0 && _cardinalityOfCatCol < (Integer.MAX_VALUE / 2)  : "Cardinality of categ. column should be within range (0, Integer.MAX_VALUE / 2 )";
+        _encodingDataPerNode = new int[maxFoldValue + 1][_cardinalityOfCatCol * 2];
+      }
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+      Chunk categoricalChunk = cs[_categoricalColumnIdx];
+      Chunk numeratorChunk = cs[_numeratorIdx];
+      Chunk denominatorChunk = cs[_denominatorIdx];
+      for (int i = 0; i < categoricalChunk.len(); i++) {
+        int levelValue = (int) categoricalChunk.at8(i);
+
+          // We are allowed to do casting to `int` as we have validation before submitting this MRTask
+          int foldValue = _foldColumnIdx != -1 ? (int) cs[_foldColumnIdx].at8(i) : 0;
+          int[] arrForNumeratorsAndDenominators = _encodingDataPerNode[foldValue];
+
+        synchronized (arrForNumeratorsAndDenominators) { // Sync on whole array as there is no way to sync on two objects ( two rows of 2D array)
+          arrForNumeratorsAndDenominators[levelValue] = (int) numeratorChunk.at8(i);
+          arrForNumeratorsAndDenominators[_cardinalityOfCatCol + levelValue] = (int) denominatorChunk.at8(i);
+        }
+      }
+    }
+
+    @Override
+    public void reduce(FrameWithEncodingDataToArray mrt) {
+      int[][] leftArr = getEncodingDataArray();
+      int[][] rightArr = mrt.getEncodingDataArray();
+      for(int rowIdx = 0 ; rowIdx < leftArr.length; rowIdx++) {
+        for(int colIdx = 0 ; colIdx < leftArr[rowIdx].length; colIdx++) {
+          synchronized (leftArr[rowIdx]) {
```

Comment:
same here, reduce is single threaded and no other thread should access `_encodingDataPerNode` at that point: all maps are completed.

Reply:
@seb-h2o Are you sure about single thread? I think it depends on how many maps we have and we can reduce in parallel as well. At least it is my current understanding.

Reply:
I was wrong, answered in previous comment.